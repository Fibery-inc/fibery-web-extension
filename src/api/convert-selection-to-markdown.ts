import rehypeParse from "rehype-parse";
import rehypeRemark from "rehype-remark";
import remarkStringify from "remark-stringify";
import { unified } from "unified";
import { visitParents } from "unist-util-visit-parents";

export async function convertSelectionToMarkdown(
  html: string,
  pageUrl: string
): Promise<string> {
  try {
    const file = await unified()
      .use(rehypeParse)
      // @ts-ignore
      .use(rehypeRemark)
      .use(() => {
        return (tree) => {
          // @ts-ignore
          visitParents(tree, ["image", "link"], (node: any) => {
            const src = (node?.url as string) || "";
            if (!src || !pageUrl) {
              return;
            }
            if (src.startsWith("http")) {
              return;
            }
            if (src.startsWith("/")) {
              const hostname = new URL(pageUrl).hostname;
              node.url = `https://${hostname}${src}`;
              return;
            }
            node.url = `${
              pageUrl.endsWith("/") ? pageUrl : pageUrl + "/"
            }${src}`;
            return;
          });
          return tree;
        };
      })
      .use(remarkStringify)
      .process(html);
    return String(file);
  } catch (error) {
    console.error("Error converting selection to markdown", error);
    return "";
  }
}
