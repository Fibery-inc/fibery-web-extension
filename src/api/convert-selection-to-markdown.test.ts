/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { expect, test } from "vitest";
import { convertSelectionToMarkdown } from "./convert-selection-to-markdown";

test("convertSelectionToMarkdown", async () => {
  const markdown = await convertSelectionToMarkdown(
    "<div>hello<p id='second'>world <a href='#inner'>test</a> <a href='/inner'>test</a> <a href='https://fibery.io'>test</a></p><img alt='' src='/img.jpg' /><br/><img alt='' src='img.jpg' /><br/><img alt='' src='https://fibery.io/img.jpg' /> ellow</div>",
    "http://fibery.io/page"
  );
  expect(markdown).toMatchSnapshot(`<p>hello</p><p>world</p>`);
});
