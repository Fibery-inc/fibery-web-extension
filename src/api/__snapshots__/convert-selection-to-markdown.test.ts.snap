// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertSelectionToMarkdown > <p>hello</p><p>world</p> 1`] = `
"hello

world [test](http://fibery.io/page/#inner) [test](https://fibery.io/inner) [test](https://fibery.io)

![](https://fibery.io/img.jpg)\\
![](http://fibery.io/page/img.jpg)\\
![](https://fibery.io/img.jpg) ellow
"
`;
